ddtrace/internal/ci_visibility/recorder.py

```diff
@@ -38,6 +41,9 @@ def __init__(self, tracer=None, config=None, service=None):
         super(CIVisibility, self).__init__()
 
         self.tracer = tracer or ddtrace.tracer
+        if ddconfig.ci_visibility_agentless_enabled and not isinstance(self.tracer._writer, CIVisibilityWriter):
+            writer = CIVisibilityWriter()
+            self.tracer.configure(writer=writer)
```

Comment:
If `trace.configure(writer=AgentWriter(...))` is called within a test the `CIVisibilityWriter` will get replaced. Is this a valid edge case?

Reply:
If someone does that in a test, I'd assume they're testing some part of the AgentWriter's behavior, which I'd say is a fine thing to allow.

Reply:
My concern here is that CIVisibility could be enabled on an application that sets `ddconfig.ci_visibility_agentless_enabled =True` and calls `tracer.configure(writer=AgentWriter(...))` to configure tracing for staging/production environments. When the code path configuring the agent writer is hit in test runs, the `CIVisibilityWriter` will be removed and all traces will be sent to a datadog agent. If a datadog agent is not running in the ci environment then all ci app spans will be dropped. 

I am not sure if this a common scenario but it might make sense to have two instances of `ddrace.Tracer()`. One used when testing an application in ci and one used in deployed applications (ex: production/staging environments). What do you think?

Reply:
I guess the part of this I don't fully understand is: why would a user set up an `AgentWriter` if they didn't also have an agent running?

Reply:
> CIVisibility could be enabled on an application that sets ddconfig.ci_visibility_agentless_enabled =True and calls tracer.configure(writer=AgentWriter(...)) to configure tracing for staging/production environments.

@mabdinur I'm unsure what you mean by this. Agentless reporting is meant to replace our AgentWriter for CIVisibility, and the latter is meant to be used only if Agentless mode is not enabled. There is no case where the user should specify using the AgentWriter in a CIVisibility context, especially since the tracer is wrapped on a private instance of the `CIVisibility` service.

Reply:
The scenario I am describing is if an application manually configures an agent writer (for example this can occur if sampling priority, sampler or statd url is updated) and then an integration test run which executes this code path.

In this scenario the agentless ci visibility writer could be replaced with a datadog agent writer. Since a datadog agent may not be running in the ci environment then spans will not be submitted. 

A fix could be to update the 'Tracer.configure()' method to check the type of the configured writer. Currently it always replaces the current writer with an agent writer.

Reply:
I think I understand. You're suggesting a change like this, right? https://github.com/DataDog/dd-trace-py/compare/emmett.butler/conf-writer